yksgi/src/lib.rs

```diff
@@ -0,0 +1,344 @@
+use llvm_sys::core::*;
+use llvm_sys::{LLVMOpcode, LLVMTypeKind};
+use std::collections::HashMap;
+use std::ffi::CStr;
+use std::ptr;
+
+mod llvmbridge;
+use llvmbridge::{LLVMModule, LLVMBasicBlock, LLVMValue, LLVMType, get_aot_original};
+
+/// Stopgap interpreter values.
+#[derive(Clone, Copy, PartialEq)]
+pub struct SGValue {
+    pub val: u64,
+    pub ty: LLVMType,
+}
+
+impl SGValue {
+    pub fn new(val: u64, ty: LLVMType) -> Self {
+        SGValue { val, ty }
+    }
+
+    pub fn with_type(&self, ty: LLVMType) -> Self {
+        SGValue::new(self.val, ty)
+    }
+
+    pub fn as_str(&self) -> String {
+        format!("SGValue ( {}, {:?} )", self.val, unsafe { self.ty.as_str() })
+    }
+}
+
+/// A frame holding live variables.
+struct Frame {
+    vars: HashMap<LLVMValue, SGValue>,
+}
+
+impl Frame {
+    fn new() -> Frame {
+        Frame {
+            vars: HashMap::new(),
+        }
+    }
+
+    /// Get the value of the variable `key` in this frame.
+    fn get(&self, key: &LLVMValue) -> Option<&SGValue> {
+        self.vars.get(key)
+    }
+
+    /// Insert new variable into this frame.
+    fn add(&mut self, key: LLVMValue, val: SGValue) {
+        if let Some(entry) = self.vars.get_mut(&key) {
+            *entry = val;
+        } else {
+            self.vars.insert(key, val);
+        }
+    }
+}
+
+/// The stopgap interpreter. Used during guard failures to get back to the control point by
+/// interpreting LLVM IR.
+pub struct SGInterp {
+    /// LLVM IR module we are interpreting.
+    module: LLVMModule,
+    /// Current frames.
+    frames: Vec<Frame>,
+    /// Current instruction being interpreted.
+    pc: LLVMValue,
+    /// Last basic block that was interpreted.
+    lastbb: Option<LLVMBasicBlock>,
```

Comment:
This looks like it owns the basic block? Should it be a reference, or a basic block index?

Reply:
I was curious. llvm-sys defines the type as an empty enum: https://docs.rs/llvm-sys/latest/llvm_sys/enum.LLVMBasicBlock.html

Is this a way of tagging an opaque pointer or something?

Reply:
This is a bit confusing, since llvm-sys has types of the same name. But these are my own types defined in `llvmbridge.rs`. So the `LLVMBasicBlock` type above is actuallly just a wrapper around an `LLVMBasicBlockRef`: `pub struct LLVMBasicBlock(pub LLVMBasicBlockRef);`

Reply:
I'm going to be rude and say that `LLVMBasicBlock` is then not a great name :) Why bother even wrapping it? Can't we just use `LLVMBasicBlockRef` as is? As a bonus, the name is better!

Reply:
Which is just a raw pointer that can be freely copied, I think?

If so, the block isn't owned.

Reply:
I'm wrapping it so that I can implement some functions on it.

Reply:
OK, so let's at least call it `LLVMBasicBlockRef` or similar.

Reply:
When I made the wrapper I didn't know `llvm-sys` has a type of the same name. If I call it `LLVMBasicBlockRef` though then it's again the same name as in `llvm-sys`. Maybe `LLVMBasicBlockRefWrapper` or `WBasicBlockRef`?

Reply:
Do we use both names in this namespace?

Reply:
We might in the case of `LLVMBasicBlockRef`, but probably not in the case of `LLVMBasicBlock`.

Reply:
If we want to use both names, then my suggestion is: don't import them and always qualify them. In other words, if two modules define `X` then don't `use A::X` or `use B::X`, but always spell out `A::X` or `B::X` in types.

Reply:
I'm currently not importing them. As I said, I didn't know `llvm-sys` even defined a `LLVMBasicBlock` type, which is why it used this name for my wrapper.

Reply:
Then why don't we just use `LLVMBasicBlockRef` for this struct?

Reply:
So just do `struct LLVMBasicBlockRef(llvm-sys::LLVMBasicBlockRef)`? I prefer the current solution over this tbh.

Reply:
> So just do struct LLVMBasicBlockRef(llvm-sys::LLVMBasicBlockRef)?

The current name is confusing (both Edd and I misunderstood it). I don't mind too much what we change it too -- I made a suggestion but there may be better alternatives!

Reply:
Fixed in [ad4e087](https://github.com/ykjit/yk/pull/500/commits/ad4e0879b40340cc9b7d404138216344872d9efd)

Reply:
That wasn't the way I was expecting... but it does work for me!